src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
+        void reset(unsigned int seedBlockNum)
+        {
+            assert(seedBlockNum <= oldBlockMaxNum);
+            oldBlocksInLoop = BlockSetOps::MakeSingleton(comp, seedBlockNum);
+            newBlocksInLoop = BlockSetOps::MakeEmpty(comp);
+        }
 
-     */
+        bool isMember(unsigned int blockNum)
+        {
+            if (blockNum > oldBlockMaxNum)
+            {
+                return BlockSetOps::IsMember(comp, newBlocksInLoop, blockNum - oldBlockMaxNum);
+            }
+            return BlockSetOps::IsMember(comp, oldBlocksInLoop, blockNum);
+        }
 
-    BasicBlock*   head;
-    BasicBlock*   top;
-    BasicBlock*   bottom;
-    BasicBlock*   entry;
-    BasicBlock*   exit;
-    unsigned char exitCount;
+        void insert(unsigned int blockNum)
+        {
+            if (blockNum > oldBlockMaxNum)
+            {
+                BlockSetOps::AddElemD(comp, newBlocksInLoop, blockNum - oldBlockMaxNum);
+            }
+            else
+            {
+                BlockSetOps::AddElemD(comp, oldBlocksInLoop, blockNum);
+            }
+        }
 
-    for (head = fgFirstBB; head->bbNext; head = head->bbNext)
+        bool testAndInsert(unsigned int blockNum)
+        {
+            if (blockNum > oldBlockMaxNum)
+            {
+                unsigned int shiftedNum = blockNum - oldBlockMaxNum;
+                if (!BlockSetOps::IsMember(comp, newBlocksInLoop, shiftedNum))
+                {
+                    BlockSetOps::AddElemD(comp, newBlocksInLoop, shiftedNum);
+                    return false;
+                }
+            }
+            else
+            {
+                if (!BlockSetOps::IsMember(comp, oldBlocksInLoop, blockNum))
+                {
+                    BlockSetOps::AddElemD(comp, oldBlocksInLoop, blockNum);
+                    return false;
+                }
+            }
+            return true;
+        }
+    };
+
+    LoopBlockSet loopBlocks; // Set of blocks identified as part of the loop
+    Compiler*    comp;
+
+    // See LoopSearch class comment header for a diagram relating these fields:
+    BasicBlock* head;   // Predecessor of unique entry edge
+    BasicBlock* first;  // Lexically first in-loop block
+    BasicBlock* top;    // Successor of back-edge from BOTTOM
+    BasicBlock* bottom; // Predecessor of back-edge to TOP, also lexically last in-loop block
+    BasicBlock* entry;  // Successor of unique entry edge
+
+    BasicBlock*   lastExit;       // Most recently discovered exit block
+    unsigned char exitCount;      // Number of discovered exit edges
+    unsigned int  oldBlockMaxNum; // Used to identify new blocks created during compaction
+    BlockSet      bottomBlocks;   // BOTTOM blocks of already-recorded loops
+#ifdef DEBUG
+    bool forgotExit = false; // Flags a rare case where lastExit gets nulled out, for assertions
+#endif
+
+public:
+    bool changedFlowGraph = false; // Signals to caller that fgUpdateChangedFlowGraph needs to be called
+
```

Comment:
Do the class users both set and read it? If the class sets it and users read it, it would be safer for the users to not have write access.